scandium-core/src/main/java/org/eclipse/californium/scandium/dtls/SignatureAlgorithmsExtension.java

```diff
@@ -0,0 +1,126 @@
+			writer.write(listLength + (LIST_LENGTH_BITS / Byte.SIZE), LENGTH_BITS);
+			writer.write(listLength, LIST_LENGTH_BITS);
+
+			for (SignatureAndHashAlgorithm signatureAndHashAlgorithm : signatureAndHashAlgorithms) {
+				writer.write(signatureAndHashAlgorithm.getHash().getCode(), SIGNATURE_BITS);
+				writer.write(signatureAndHashAlgorithm.getSignature().getCode(), HASH_BITS);
+			}
+		}
+
+		public static HelloExtension fromExtensionDataReader(DatagramReader extensionDataReader) {
+
+			List<SignatureAndHashAlgorithm> signatureAndHashAlgorithms = new ArrayList<SignatureAndHashAlgorithm>();
+			int listLength = extensionDataReader.read(LIST_LENGTH_BITS);
+			DatagramReader rangeReader = extensionDataReader.createRangeReader(listLength);
+			while (rangeReader.bytesAvailable()) {
+				int hashId = rangeReader.read(SIGNATURE_BITS);
+				int signatureId = rangeReader.read(HASH_BITS);
+
+				signatureAndHashAlgorithms.add(new SignatureAndHashAlgorithm(hashId, signatureId));
+			}
+			SignatureAndHashAlgorithm[] signatureAndHashAlgoritms = signatureAndHashAlgorithms.toArray(new SignatureAndHashAlgorithm[] {});
+			return new SignatureAlgorithmsExtension(signatureAndHashAlgoritms);
+		}
+		
+		// Methods ////////////////////////////////////////////////////////
+		
+		@Override
+		public int getLength() {
+			// fixed: type (2 bytes), length (2 bytes), list length (2 bytes)
+			// variable: number of signature algorithms * 2 (1 byte for signature algorithm, 1 byte for hash algorithm )
+			return 6 + (signatureAndHashAlgorithms.size() * (SIGNATURE_ALGORITHM_BITS / Byte.SIZE));
+		}
+
+		@Override
+		public String toString() {
+			StringBuilder sb = new StringBuilder(super.toString());
+			sb.append("\t\t\t\tLength: ").append(getLength() - 4);
+			sb.append(StringUtil.lineSeparator()).append("\t\t\t\tSignature Algorithms Length: ").append(getLength() - 6);
+			sb.append(StringUtil.lineSeparator()).append("\t\t\t\tSignature Algorithms (").append(signatureAndHashAlgorithms.size()).append(" algorithm):");
+
+			for (SignatureAndHashAlgorithm signatureAndHashAlgoritm : signatureAndHashAlgorithms) {
+				sb.append(StringUtil.lineSeparator()).append("\t\t\t\t\tSignature Algorithm: ");
+				sb.append(StringUtil.lineSeparator()).append("\t\t\t\t\t\tSignature Hash Algorithm Hash: ");
+				if (signatureAndHashAlgoritm.getHash() != null) {
+					sb.append(signatureAndHashAlgoritm.getHash());
+				} else {
+					sb.append("unknown");
+				}
+				sb.append(StringUtil.lineSeparator()).append("\t\t\t\t\t\tSignature Hash Algorithm Signature: ");
+				if (signatureAndHashAlgoritm.getSignature() != null) {
+					sb.append(signatureAndHashAlgoritm.getSignature());
+				} else {
+					sb.append("unknown");
+				}
+			}
+			sb.append(StringUtil.lineSeparator());
+			return sb.toString();
+		}
+
+		public List<SignatureAndHashAlgorithm> getSupportedSignatureAndHashAlgorithmIds() {
+			return Collections.unmodifiableList(signatureAndHashAlgorithms);
+		}
+}
```

Comment:
If you use the Eclipse IDE, the project contains a "eclipse-formatter-profile.xml", please apply that.
If you don't use the Eclipse IDE, I apply the formatter in a separate commit after merging.

Reply:
I applied the formatter.